deploy-agent/deployd/common/stats.py

```diff
@@ -28,69 +49,287 @@ def __exit__(self, stats, sample_rate=1.0, tags=None):
+            return: bool
+        """
+        return not os.stat(self.path).st_size > 0
+
+    def read(self):
+        """ read metrics from cache, then delete
+            return: generator
+        """
+        with open(self.path, 'r') as fh:
+            for line in fh:
+                yield Stat(ins=line)
+
+    def write(self, output):
+        """ write metrics to cache file respecting max cache size
+            appends newline to metric
+        """
+        if self.limit():
+            msg = 'cache file {} has exceeded maximum file size of {}'
+            log.error(msg.format(self.path, self.max_size))
+            return
+        with open(self.path, 'a') as fh:
+            fh.write('{}\n'.format(output))
+
+    def truncate(self):
+        """ purge cache file """
+        with open(self.path, 'w') as fh:
+            fh.truncate()
+
+
+class Stat:
+    """ stat class for simple data management, dataclasses are py3.7+
+        supports all methods for stats
+    """
+
+    def __init__(self, mtype=None, name=None, value=None, sample_rate=None, tags=None, ins=None):
+        self.mtype = mtype
+        self.name = name
+        self.value = value
+        self.sample_rate = sample_rate
+        self.tags = tags
+        self.ins = ins
+        try:
+            self.JSONDecodeError = json.decoder.JSONDecodeError
+        except AttributeError:
+            # python2 support
+            self.JSONDecodeError = ValueError
+
+    def __repr__(self):
+        obj = dict()
+        obj['mtype'] = self.mtype
+        if self.value is not None:
+            # value can be 0
+            obj['value'] = self.value
+        obj['name'] = self.name
+        obj['sample_rate'] = self.sample_rate
+        obj['tags'] = self.tags
+        return obj
+
+    def serialize(self):
+        """ serialize for cache writing """
+        return json.dumps(self.__repr__())
+
+    def _deserialize(self):
+        """ read in json, setting defaults for a stat
+            return: None
+        """
+        obj = json.loads(self.ins)
+        self.mtype = obj.get('mtype', None)
+        self.name = obj.get('name', None)
+        self.value = obj.get('value', None)
+        self.sample_rate = obj.get('sample_rate', None)
+        self.tags = obj.get('tags', None)
+
+    def deserialize(self, ins=None):
+        """ attempt to deserialize
+            :param: ins json as str
+            return: bool, False on error
+        """
+        if ins:
+            self.ins = ins
+        try:
+            self._deserialize()
+            return True
+        except self.JSONDecodeError:
+            return False
+        except TypeError:
+            return False
+
+
+class MetricClient:
+    """ metrics client wrapper, enables disk cache """
+
+    def __init__(self, port=METRIC_PORT_HEALTH, cache_path=METRIC_CACHE_PATH):
+        self.port = port
+        self.cache = MetricCache(path=cache_path)
+        self.stat = None
```

Comment:
what's the use of stat if it's lifetype is limited to _send()

Reply:
Stat is used by MetricClient and MetricCache (MetricClient.parse_stat and MetricCache.read) and this member variable is used by both flushing cache, reading, and sending metrics.